docs/poly-to-normal-variant/index.md

```diff
@@ -0,0 +1,345 @@
+all, but to be explicit, we can force it to be a polymorphic variant:
+
+<<< Types.re#must-be-poly{4}
+
+The `[> ]` type expression means a polymorphic variant that has no tags, but
+allows more tags, which basically means any polymorphic variant. Note that
+adding this small restriction to the type doesn't make a real difference in this
+program---it's just a way to make it clear that `DiscountError`'s argument
+should be a polymorphic variant. It's an optional embellishment that you can
+feel free to leave out.
+
+---
+
+Phew! You refactored the `discount` derived variable to use a normal variant
+instead of a polymorphic variant. The code changes were fairly minimal, but to
+understand what was happening, it was necessary to learn the basics of type
+constructors and type variables. In the next sections, we'll set types and other
+theoretical considerations aside and get into the nitty-gritty of the UI changes
+you must make to add promo support to the `Order` component.
+
+## Overview
+
+- A type constructor takes a type and outputs another type
+- A type variable is a variable that stands in a for a type and often appears in
+  type constructors or type signatures
+- In polymorphic variant type expressions, `>` means that the polymorphic
+  variant can accept more than the variant tags that are listed
+  - You rarely need to use `>` in your own type definitions, but it often
+    appears in inferred type definitions (that appear when you hover over
+    variables and functions)
+  - Inferred type definitions that contain `>` also have an implicit type
+    variable
+
+## Exercises
+
+<b>1.</b> The following code
+([playground](https://melange.re/v4.0.0/playground/?language=Reason&code=bGV0IGdldE5hbWUgPSAoYW5pbWFsOiBbfCBgQ2F0IHwgYERvZyhpbnQpIHwgYFVuaWNvcm4oc3RyaW5nKV0pID0%2BCiAgc3dpdGNoIChhbmltYWwpIHsKICB8IGBDYXQgPT4gIk1yIFdoaXNrZXJzIgogIHwgYERvZyhuKSA9PiAiQmFuZGl0ICIgKysgc3RyaW5nX29mX2ludChuKQogIHwgYFVuaWNvcm4obmFtZSkgPT4gIlNpciAiICsrIG5hbWUKICB9Ow%3D%3D&live=off))
+doesn’t compile. Fix it by adding a single character.
+
+```reason
+let getName = (animal: [| `Cat | `Dog(int) | `Unicorn(string)]) =>
+  switch (animal) {
+  | `Cat => "Mr Whiskers"
+  | `Dog(n) => "Bandit " ++ string_of_int(n)
+  | `Unicorn(name) => "Sir " ++ name
+  };
+```
+
+::: details Hint
+
+Find a place where you can insert a space.
+
+:::
+
+::: details Solution
+
+```reason{1}
+let getName = (animal: [ | `Cat | `Dog(int) | `Unicorn(string)]) =>
+  switch (animal) {
+  | `Cat => "Mr Whiskers"
+  | `Dog(n) => "Bandit " ++ string_of_int(n)
+  | `Unicorn(name) => "Sir " ++ name
+  };
+```
+
+A common mistake when writing polymorphic variant type definitions is forgetting
+to put a space between the `[` and the `|` characters. Note that you don't need
+to add the implicit type variable in type annotations.
+
+::: warning
+
+In the next version of Melange, polymorphic variant definitions no longer
+require a space between `[` and `|`.
```

Comment:
Wait :wat: is this because of some changes in the OCaml compiler, or Reason? In any case, I don't think this is Melange responsibility, so I would not add this warning unless it refers to the last package and version that didn't support it and the first one that will.

Reply:
Then we shouldn't refer to Melange new version. Reason is independent package than Melange, and they can be combined in different ways. In other words, having Melange 4.0.1 doesn't involve getting the latest reason.